### PR TITLE
Feat/met 4451 integration of problem patterns

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: src/test/java/eu/europeana/metis/sandbox/service/problempatterns/PatternAnalysisServiceImplTest.java [hunk 1]
```diff
@@ -203,14 +203,12 @@ void getRecordPatternAnalysisTest() {
     final LocalDateTime nowP2 = LocalDateTime.now();
     patternAnalysisService.generateRecordPatternAnalysis("1", Step.VALIDATE_INTERNAL, nowP2, rdfRecordP2);
     //We should be getting this from the database
-    List<ProblemPattern> problemPatternsRecord1 = patternAnalysisService.getRecordPatternAnalysis("1",
-        Step.VALIDATE_INTERNAL, nowP2, rdfRecordP2);
+    List<ProblemPattern> problemPatternsRecord1 = patternAnalysisService.getRecordPatternAnalysis(rdfRecordP2);
     assertFalse(problemPatternsRecord1.isEmpty());
 
     final LocalDateTime nowP6 = LocalDateTime.now();
```
Comment (jeortizquan) [inline src/test/java/eu/europeana/metis/sandbox/service/problempatterns/PatternAnalysisServiceImplTest.java:209]: this variable is not used.

**Review by jeortizquan — CHANGES_REQUESTED** [review]
There are some comments to address. In general looks fine.

File: src/main/java/eu/europeana/metis/sandbox/controller/PatternAnalysisController.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        private final String datasetId;
+        @JsonProperty
+        private final T executionStep;
+        @JsonProperty
+        private final String executionTimestamp;
+        @JsonProperty
+        private final List<ProblemPattern> problemPatternList;
+
+        private DatasetProblemPatternAnalysisView( DatasetProblemPatternAnalysis<T> datasetProblemPatternAnalysis) {
+            this.datasetId = datasetProblemPatternAnalysis.getDatasetId();
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/PatternAnalysisController.java:113]: Just ran the intellij formatter, hopefully now is better but I didn't see much change

File: src/main/java/eu/europeana/metis/sandbox/controller/advice/ControllerErrorHandler.java [hunk 3]
```diff
@@ -80,6 +83,24 @@ public ResponseEntity<Object> handleInvalidDatasetException(
     return new ResponseEntity<>(exceptionModel, exceptionModel.getStatus());
   }
 
+  @ExceptionHandler(SerializationException.class)
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/controller/advice/ControllerErrorHandler.java:86]: Just ran the formatter, hopefully now is better

File: src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java [hunk 4]
```diff
@@ -30,6 +34,10 @@ public InternalValidationExecutor(AmqpTemplate amqpTemplate,
       containerFactory = "internalValidationFactory",
       autoStartup = "${sandbox.rabbitmq.queues.record.transformed.auto-start:true}")
   public void validateInternal(RecordProcessEvent input) {
-    consume(routingKey, input, Step.VALIDATE_INTERNAL, () -> service.validate(input.getRecord()));
+    if(!hasFirstRecordArrived) {
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java:37]: Good thing you brought this up, this was clearly not thought through all the way...
The idea is to detect when the step validation internal is first used for a certain dataset so we can save in the database.
The way this is written right now will not execute this idea for every dataset. Need to rethink about this

[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java:37]: I tried using the database, as in detecting if there was already an entry in the database, and it didn't work quite well, so I tried something different and is working now

File: src/test/java/eu/europeana/metis/sandbox/service/problempatterns/PatternAnalysisServiceImplTest.java (hunk 1 above)
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/problempatterns/PatternAnalysisServiceImplTest.java:209]: Removed

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java [hunk 5]
```diff
@@ -2,35 +2,52 @@
...
           validationResult.getRecordId(), validationResult.getNodeId());
     }
-
-    return new RecordInfo(record);
+    try {
+      patternAnalysisService.generateRecordPatternAnalysis(recordToValidate.getDatasetId(), Step.VALIDATE_INTERNAL, timestamp,
+              new String(recordToValidate.getContent(), StandardCharsets.UTF_8));
+    } catch (PatternAnalysisException e) {
+      LOGGER.error(String.format("An error occurred while processing pattern analysis with record id %s", recordToValidate.getEuropeanaId()));
+    }
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java:50]: Done

File: src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java [hunk 6]
```diff
@@ -19,6 +21,8 @@ class InternalValidationExecutor extends StepExecutor {
 
   @Value("${sandbox.rabbitmq.queues.record.validated.internal.queue}")
   private String routingKey;
+  private boolean hasFirstRecordArrived = false;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java:24]: this variable does not exist anymore

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java [hunk 7]
```diff
@@ -2,35 +2,63 @@
...
+        } catch (PatternAnalysisException e) {
+            LOGGER.error("An error occurred while processing pattern analysis with record id {}", recordToValidate.getEuropeanaId());
+        }
+        return new RecordInfo(recordToValidate);
     }
 
-    return new RecordInfo(record);
-  }
+    private void generateAnalysis(String datasetId, byte[] recordContent) throws PatternAnalysisException {
+        Optional<ExecutionPoint> firstOccurrence = executionPointService.getExecutionPoint(datasetId, Step.VALIDATE_INTERNAL.toString());
```
Comment (jeortizquan) [inline src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java:59]: We checked together with Joana this method. we will continue to solve the sync with MET-4470

**Review by jeortizquan — APPROVED** [review]
Now the occurrences are stored in different rows. The synchronization solution should address this issue and it will be attended in MET-4470.

File: src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/executor/workflow/InternalValidationExecutor.java]: We discussed this topic. taking care in MET-4470.

**Pull request merged**